File: src/sc_neurocore/accel/go/services/sc_chaotic_map_neuron.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Go service mirror for the SC two-state chaotic map

package services

import (
	"errors"
	"math"
)

// SCChaoticMapNeuronState holds the two project-defined map states.
type SCChaoticMapNeuronState struct {
	X          float64
	Y          float64
	KF         float64
	KS         float64
	Alpha      float64
	Delta      float64
	XThreshold float64
}

// NewSCChaoticMapNeuron creates the preserved SC engineering model.
func NewSCChaoticMapNeuron() *SCChaoticMapNeuronState {
	return &SCChaoticMapNeuronState{
		X: 0.0, Y: 0.0, KF: 0.7, KS: 0.95, Alpha: 2.0, Delta: 0.05, XThreshold: 0.5,
	}
}

func finiteSCChaoticMap(value float64) bool {
	return !math.IsNaN(value) && !math.IsInf(value, 0)
}

func logisticSCChaoticMap(z float64) float64 {
	if z >= 0.0 {
		return 1.0 / (1.0 + math.Exp(-z))
	}
	expZ := math.Exp(z)
	return expZ / (1.0 + expZ)
}

// ValidateSCChaoticMap checks state and numerical parameters.
func ValidateSCChaoticMap(s *SCChaoticMapNeuronState) bool {
	if s == nil {
		return false
	}
	return finiteSCChaoticMap(s.X) && finiteSCChaoticMap(s.Y) &&
		finiteSCChaoticMap(s.KF) && s.KF >= 0.0 &&
		finiteSCChaoticMap(s.KS) && finiteSCChaoticMap(s.Alpha) &&
		finiteSCChaoticMap(s.Delta) && s.Delta >= 0.0 &&
		finiteSCChaoticMap(s.XThreshold)
}

// Step advances both states simultaneously and returns an upward crossing.
func (s *SCChaoticMapNeuronState) Step(current float64) (int, error) {
	if !ValidateSCChaoticMap(s) {
		return 0, errors.New("invalid SC chaotic map runtime state")
	}
	if !finiteSCChaoticMap(current) {
		return 0, errors.New("invalid SC chaotic map current")
	}

	xPrev := s.X
	xNew := s.KF*s.X*logisticSCChaoticMap(s.X+s.Alpha) - s.Y + current
	yNew := s.KS*s.Y + s.Delta*s.X
	if !finiteSCChaoticMap(xNew) || !finiteSCChaoticMap(yNew) {
		return 0, errors.New("invalid SC chaotic map candidate state")
	}
	s.X = math.Min(10.0, math.Max(-10.0, xNew))
	s.Y = math.Min(10.0, math.Max(-10.0, yNew))
	if s.X >= s.XThreshold && xPrev < s.XThreshold {
		return 1, nil
	}
	return 0, nil
}

// SimulateSCChaoticMapNeuron runs the map for n steps.
func SimulateSCChaoticMapNeuron(nSteps int, current float64) ([]float64, int) {
	s := NewSCChaoticMapNeuron()
	trace := make([]float64, nSteps)
	spikes := 0
	for t := 0; t < nSteps; t++ {
		result, err := s.Step(current)
		if err != nil {
			trace[t] = math.NaN()
			continue
		}
		trace[t] = s.X
		spikes += result
	}
	return trace, spikes
}

File: src/sc_neurocore/accel/julia/neurons/sc_chaotic_map_neuron.jl
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Julia mirror for the SC two-state chaotic map

module SCChaoticMapNeuronAccel

export step!, simulate, simulate_sc_chaotic_map, validate, SCChaoticMapNeuronState

mutable struct SCChaoticMapNeuronState
    x::Float64
    y::Float64
    k_f::Float64
    k_s::Float64
    alpha::Float64
    delta::Float64
    x_threshold::Float64
end

function SCChaoticMapNeuronState()
    SCChaoticMapNeuronState(0.0, 0.0, 0.7, 0.95, 2.0, 0.05, 0.5)
end

function validate(s::SCChaoticMapNeuronState)::Bool
    return isfinite(s.x) &&
        isfinite(s.y) &&
        isfinite(s.k_f) &&
        s.k_f >= 0.0 &&
        isfinite(s.k_s) &&
        isfinite(s.alpha) &&
        isfinite(s.delta) &&
        s.delta >= 0.0 &&
        isfinite(s.x_threshold)
end

function logistic(z::Float64)::Float64
    if z >= 0.0
        return 1.0 / (1.0 + exp(-z))
    end
    exp_z = exp(z)
    return exp_z / (1.0 + exp_z)
end

function step!(s::SCChaoticMapNeuronState, current::Float64=0.0; dt::Float64=0.1)
    if !validate(s) || !isfinite(current) || !isfinite(dt)
        return -1
    end

    x_prev = s.x
    x_new = s.k_f * s.x * logistic(s.x + s.alpha) - s.y + current
    y_new = s.k_s * s.y + s.delta * s.x
    if !isfinite(x_new) || !isfinite(y_new)
        return -1
    end
    s.x = clamp(x_new, -10.0, 10.0)
    s.y = clamp(y_new, -10.0, 10.0)
    return (s.x >= s.x_threshold && x_prev < s.x_threshold) ? 1 : 0
end

function simulate(n_steps::Int=1000; current::Float64=10.0, dt::Float64=0.1)
    s = SCChaoticMapNeuronState()
    trace = zeros(n_steps)
    spikes = 0
    for t in 1:n_steps
        result = step!(s, current; dt=dt)
        trace[t] = s.x
        if result > 0
            spikes += 1
        end
    end
    return trace, spikes
end

function simulate_sc_chaotic_map(
    x::Real,
    y::Real,
    k_f::Real,
    k_s::Real,
    alpha::Real,
    delta::Real,
    x_threshold::Real,
    current::AbstractVector,
)
    state = SCChaoticMapNeuronState(Float64.((x, y, k_f, k_s, alpha, delta, x_threshold))...)
    validate(state) || throw(ArgumentError("invalid SC chaotic-map configuration"))
    all(isfinite, current) || throw(ArgumentError("current must contain only finite values"))
    steps = length(current)
    x_trace = Vector{Float64}(undef, steps)
    y_trace = Vector{Float64}(undef, steps)
    spikes = Vector{Float64}(undef, steps)
    count = 0
    @inbounds for index in 1:steps
        event = step!(state, Float64(current[index]))
        event >= 0 || throw(ArgumentError("invalid SC chaotic-map candidate"))
        x_trace[index] = state.x
        y_trace[index] = state.y
        spikes[index] = Float64(event)
        count += event
    end
    return (x_trace, y_trace, spikes, state.x, state.y, count)
end

end # module SCChaoticMapNeuronAccel
